File: src/java/com/kitware/pulse/cdm/system/equipment/ecmo/SEECMOSettings.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.pulse.cdm.system.equipment.ecmo;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import com.kitware.pulse.cdm.bind.Enums.eSwitch;
import com.kitware.pulse.cdm.bind.ECMO.ECMOSettingsData;
import com.kitware.pulse.cdm.bind.ECMO.eECMOCannulationLocation;
import com.kitware.pulse.cdm.bind.Substance.SubstanceConcentrationData;
import com.kitware.pulse.cdm.bind.Substance.SubstanceFractionData;
import com.kitware.pulse.cdm.properties.*;
import com.kitware.pulse.cdm.substance.SESubstanceConcentration;
import com.kitware.pulse.cdm.substance.SESubstanceFraction;

public class SEECMOSettings
{
  protected eECMOCannulationLocation          inflowLocation;
  protected eECMOCannulationLocation          outflowLocation;
  protected SEScalarVolume                    oxygenatorVolume;
  protected SEScalarVolumePerTime             transfusionFlow;
  protected String                            substanceCompound;
  
  protected List<SESubstanceConcentration>    substanceConcentrations;


  public SEECMOSettings()
  {
    inflowLocation = null;
    outflowLocation = null;
    oxygenatorVolume = null;
    transfusionFlow = null;
    substanceCompound = null;

    this.substanceConcentrations=new ArrayList<SESubstanceConcentration>();
  }

  public void clear()
  {
    inflowLocation = null;
    outflowLocation = null;
    if (oxygenatorVolume != null)
      oxygenatorVolume.invalidate();
    if (transfusionFlow != null)
      transfusionFlow.invalidate();
    substanceCompound = null;
    
    this.substanceConcentrations.clear();
  }

  public void copy(SEECMOSettings from)
  {
    clear();
    if(from.inflowLocation!=null && from.inflowLocation != eECMOCannulationLocation.UNRECOGNIZED)
    	this.inflowLocation=from.inflowLocation;
    if(from.outflowLocation!=null && from.outflowLocation != eECMOCannulationLocation.UNRECOGNIZED)
      this.outflowLocation=from.outflowLocation;
    if(from.hasOxygenatorVolume())
      this.getOxygenatorVolume().set(from.getOxygenatorVolume());
    if(from.hasTransfusionFlow())
      this.getTransfusionFlow().set(from.getTransfusionFlow());
    substanceCompound = from.substanceCompound;
    
    if(from.substanceConcentrations!=null)
    {
      SESubstanceConcentration mine;
      for(SESubstanceConcentration sc : from.substanceConcentrations)
      {
        mine=this.createSubstanceConcentration(sc.getSubstance());
        if(sc.hasConcentration())
          mine.getConcentration().set(sc.getConcentration());
      }
    }
  }

  public static void load(ECMOSettingsData src, SEECMOSettings dst)
  {
    dst.clear();
    if (src.getInflowLocation()!=eECMOCannulationLocation.UNRECOGNIZED)
      dst.setInflowLocation(src.getInflowLocation());
    if (src.getOutflowLocation()!=eECMOCannulationLocation.UNRECOGNIZED)
      dst.setOutflowLocation(src.getOutflowLocation());
    if (src.hasOxygenatorVolume())
      SEScalarVolume.load(src.getOxygenatorVolume(), dst.getOxygenatorVolume());
    if (src.hasTransfusionFlow())
      SEScalarVolumePerTime.load(src.getTransfusionFlow(), dst.getTransfusionFlow());
    dst.substanceCompound = src.getSubstanceCompound();
    
    if(src.getSubstanceConcentrationsList()!=null)
    {
      for(SubstanceConcentrationData subData : src.getSubstanceConcentrationsList())
      {
        SEScalarMassPerVolume.load(subData.getConcentration(),dst.createSubstanceConcentration(subData.getName()).getConcentration());
      }
    }
  }
  public static ECMOSettingsData unload(SEECMOSettings src)
  {
    ECMOSettingsData.Builder dst = ECMOSettingsData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEECMOSettings src, ECMOSettingsData.Builder dst)
  {
    if (src.hasInflowLocation())
      dst.setInflowLocation(src.getInflowLocation());
    if (src.hasOutflowLocation())
      dst.setOutflowLocation(src.getOutflowLocation());
    if(src.hasOxygenatorVolume())
      dst.setOxygenatorVolume(SEScalarVolume.unload(src.getOxygenatorVolume()));
    if(src.hasTransfusionFlow())
      dst.setTransfusionFlow(SEScalarVolumePerTime.unload(src.getTransfusionFlow()));
    dst.setSubstanceCompound(src.substanceCompound);
    
    for(SESubstanceConcentration sc : src.substanceConcentrations)
      dst.addSubstanceConcentrations(SESubstanceConcentration.unload(sc));
  }
  
  public eECMOCannulationLocation getInflowLocation()
  {
    return inflowLocation;
  }
  public void setInflowLocation(eECMOCannulationLocation c)
  {
    inflowLocation = (c == eECMOCannulationLocation.UNRECOGNIZED) ? null : c;
  }
  public boolean hasInflowLocation()
  {
    return inflowLocation != null && inflowLocation != eECMOCannulationLocation.UNRECOGNIZED;
  }
  
  public eECMOCannulationLocation getOutflowLocation()
  {
    return outflowLocation;
  }
  public void setOutflowLocation(eECMOCannulationLocation c)
  {
    outflowLocation = (c == eECMOCannulationLocation.UNRECOGNIZED) ? null : c;
  }
  public boolean hasOutflowLocation()
  {
    return outflowLocation != null && outflowLocation != eECMOCannulationLocation.UNRECOGNIZED;
  }
  
  public SEScalarVolume getOxygenatorVolume()
  {
    if (oxygenatorVolume == null)
      oxygenatorVolume = new SEScalarVolume();
    return oxygenatorVolume;
  }
  public boolean hasOxygenatorVolume()
  {
    return oxygenatorVolume == null ? false : oxygenatorVolume.isValid();
  }
  
  public SEScalarVolumePerTime getTransfusionFlow()
  {
    if (transfusionFlow == null)
      transfusionFlow = new SEScalarVolumePerTime();
    return transfusionFlow;
  }
  public boolean hasTransfusionFlow()
  {
    return transfusionFlow == null ? false : transfusionFlow.isValid();
  }
  
  public String getSubstanceCompound()
  {
    return substanceCompound;
  }
  public void setSubstanceCompound(String c)
  {
    substanceCompound = c;
  }
  public boolean hasSubstanceCompound()
  {
    return substanceCompound != null;
  }

  public SESubstanceConcentration createSubstanceConcentration(String substance)
  {
    return getSubstanceConcentration(substance);
  }
  public SESubstanceConcentration getSubstanceConcentration(String substance)
  {
    for(SESubstanceConcentration sc : this.substanceConcentrations)
    {
      if(sc.getSubstance().equals(substance))
      {
        return sc;
      }
    }    
    SESubstanceConcentration sc = new SESubstanceConcentration(substance);
    this.substanceConcentrations.add(sc);
    return sc;
  }
  public boolean hasSubstanceConcentration()
  {
    return !this.substanceConcentrations.isEmpty();
  }
  public boolean hasSubstanceConcentration(String substance)
  {
    for(SESubstanceConcentration sc : this.substanceConcentrations)
    {
      if(sc.getSubstance().equals(substance))
      {
        return true;
      }
    }
    return false;
  }
  public List<SESubstanceConcentration> getSubstanceConcentration()
  {
    return Collections.unmodifiableList(this.substanceConcentrations);
  }
  public void removeSubstanceConcentration(String substance)
  {
    for(SESubstanceConcentration sc : this.substanceConcentrations)
    {
      if(sc.getSubstance().equals(substance))
      {
        this.substanceConcentrations.remove(sc);
        return;
      }
    }  
  }

  @Override
  public String toString()
  {
    String str = "ECMO Settings"
        + "\n\tInflowLocation: " + (hasInflowLocation() ? getInflowLocation() : "NotProvided")
        + "\n\tOutflowLocation: " + (hasOutflowLocation() ? getOutflowLocation() : "NotProvided")
        + "\n\tOxygenatorVolume: " + (hasOxygenatorVolume() ? getOxygenatorVolume().toString() : "Not Provided")
        + "\n\tTransfusionFlow: " + (hasTransfusionFlow() ? getTransfusionFlow().toString() : "Not Provided")
        + "\n\tSubstanceCompound: " + (hasSubstanceCompound() ? getSubstanceCompound().toString() : "Not Provided");

    for(SESubstanceConcentration sc : this.substanceConcentrations)
    str += "\n\t"+sc.toString();
    return str;
  }
}

File: src/csharp/pulse/cdm/system/equipment/ecmo/SEECMOSettings.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System.Collections.Generic;

namespace Pulse.CDM
{
  // Keep enums in sync with appropriate schema/cdm/ECMO.proto file !!
  public enum eECMO_CannulationLocation : int
  {
    NullCannulationLocation = 0,
    InternalJugular,
    RightFemoralVein,
    LeftFemoralVein,
    RightSubclavianVein,
    LeftSubclavianVein
  }

  public class SEECMOSettings : SEEquipment
  {
    protected eECMO_CannulationLocation      inflow_location;
    protected eECMO_CannulationLocation      outflow_location;
    protected SEScalarVolume                 oxygenator_volume;
    protected SEScalarVolumePerTime          transfusion_flow;
    protected string                         substance_compound;
    protected List<SESubstanceConcentration> substance_concentrations;

    public SEECMOSettings()
    {
      inflow_location = eECMO_CannulationLocation.NullCannulationLocation;
      outflow_location = eECMO_CannulationLocation.NullCannulationLocation;
      oxygenator_volume = null;
      transfusion_flow = null;
      substance_compound = null;
      this.substance_concentrations = new List<SESubstanceConcentration>();
    }

    public override void Clear()
    {
      base.Clear();
      inflow_location = eECMO_CannulationLocation.NullCannulationLocation;
      outflow_location = eECMO_CannulationLocation.NullCannulationLocation;
      if (oxygenator_volume != null)
        oxygenator_volume.Invalidate();
      if (transfusion_flow != null)
        transfusion_flow.Invalidate();
      substance_compound = null;
      substance_concentrations.Clear();
    }

    public void Copy(SEECMOSettings from)
    {
      base.Copy(from);
      if (from.inflow_location != eECMO_CannulationLocation.NullCannulationLocation)
        this.inflow_location = from.inflow_location;
      if (from.outflow_location != eECMO_CannulationLocation.NullCannulationLocation)
        this.outflow_location = from.inflow_location;
      if (from.HasOxygenatorVolume())
        this.GetOxygenatorVolume().Set(from.GetOxygenatorVolume());
      if (from.HasTransfusionFlow())
        this.GetTransfusionFlow().Set(from.GetTransfusionFlow());
      substance_compound = from.substance_compound;

      if (from.substance_concentrations != null)
      {
        SESubstanceConcentration mine;
        foreach (SESubstanceConcentration sc in from.substance_concentrations)
        {
          mine = this.CreateSubstanceConcentration(sc.GetSubstance());
          if (sc.HasConcentration())
            mine.GetConcentration().Set(sc.GetConcentration());
        }
      }
    }

    public eECMO_CannulationLocation GetInflowLocation()
    {
      return inflow_location;
    }
    public void SetInflowLocation(eECMO_CannulationLocation c)
    {
      inflow_location = c;
    }
    public bool HasInflowLocation()
    {
      return inflow_location != eECMO_CannulationLocation.NullCannulationLocation;
    }

    public eECMO_CannulationLocation GetOutflowLocation()
    {
      return outflow_location;
    }
    public void SetOutflowLocation(eECMO_CannulationLocation c)
    {
      outflow_location = c;
    }
    public bool HasOutflowLocation()
    {
      return outflow_location != eECMO_CannulationLocation.NullCannulationLocation;
    }

    public SEScalarVolume GetOxygenatorVolume()
    {
      if (oxygenator_volume == null)
        oxygenator_volume = new SEScalarVolume();
      return oxygenator_volume;
    }
    public bool HasOxygenatorVolume()
    {
      return oxygenator_volume == null ? false : oxygenator_volume.IsValid();
    }

    public SEScalarVolumePerTime GetTransfusionFlow()
    {
      if (transfusion_flow == null)
        transfusion_flow = new SEScalarVolumePerTime();
      return transfusion_flow;
    }
    public bool HasTransfusionFlow()
    {
      return transfusion_flow == null ? false : transfusion_flow.IsValid();
    }

    public string GetSubstanceCompound()
    {
      return substance_compound;
    }
    public void SetSubstanceCompound(string c)
    {
      substance_compound = c;
    }
    public bool HasSubstanceCompound()
    {
      return !string.IsNullOrEmpty(substance_compound);
    }

    public SESubstanceConcentration CreateSubstanceConcentration(string substance)
    {
      return GetSubstanceConcentration(substance);
    }
    public SESubstanceConcentration GetSubstanceConcentration(string substance)
    {
      foreach (SESubstanceConcentration sc in substance_concentrations)
      {
        if (sc.GetSubstance() == substance)
        {
          return sc;
        }
      }
      SESubstanceConcentration nsc = new SESubstanceConcentration(substance);
      substance_concentrations.Add(nsc);
      return nsc;
    }
    public bool HasSubstanceConcentration()
    {
      return substance_concentrations.Count != 0;
    }
    public bool HasSubstanceConcentration(string substance)
    {
      foreach (SESubstanceConcentration sc in substance_concentrations)
      {
        if (sc.GetSubstance() == substance)
        {
          return true;
        }
      }
      return false;
    }
    public List<SESubstanceConcentration> GetSubstanceConcentrations()
    {
      return substance_concentrations;
    }
    public void RemoveSubstanceConcentration(string substance)
    {
      foreach (SESubstanceConcentration sc in substance_concentrations)
      {
        if (sc.GetSubstance() == substance)
        {
          substance_concentrations.Remove(sc);
          return;
        }
      }
    }
    public void RemoveSubstanceConcentrations()
    {
      substance_concentrations.Clear();
    }

    public override string ToString()
    {
      string str = "ECMO Settings"
      + "\n\tInflowLocation: " + (HasInflowLocation() ? eECMO.CannulationLocation_Name(GetInflowLocation()) : "NotProvided")
      + "\n\tOutflowLocation: " + (HasOutflowLocation() ? eECMO.CannulationLocation_Name(GetOutflowLocation()) : "NotProvided")
      + "\n\tOxygenatorVolume: " + (HasOxygenatorVolume() ? GetOxygenatorVolume().ToString() : "Not Provided")
      + "\n\tTransfusionFlow: " + (HasTransfusionFlow() ? GetTransfusionFlow().ToString() : "Not Provided")
      + "\n\tSubstanceCompound: " + (HasSubstanceCompound() ? GetSubstanceCompound().ToString() : "Not Provided");

      foreach (SESubstanceConcentration sc in this.substance_concentrations)
        str += "\n\t" + sc.GetSubstance();

      return str;
    }
  }
}